### PR TITLE
Features2d/draw matches

### PR DESCRIPTION
**Pull request by shanduur**
Go wrapper around native OpenCV `drawMatches` function.
- Draws the found matches of keypoints from two images.

### CODE REVIEW

**Comment by shanduur** [comment]
I think it is usable now, however I am not sure if the outcome looks as it should. I hope someone is able to confirm it is done as it should, and maybe accept the PR.

Here Outcome using OpenCV's `box.png` and `box_in_scene.png` images.

Screenshot from 2020-07-31 14-15-03

Edit: Fixed misplaced Query and Train.

**Comment by shanduur** [comment]
At this point this is finalized, ready to merge. I hope someone will take a look at it in nearest future.

Here is output of new CMD:
Screenshot from 2020-08-04 19-38-52

**Comment by sknick** [comment]
Any status on this?

**Comment by shanduur** [comment]
> Any status on this?

Waiting for approval.

**Comment by shanduur** [comment]
@JustInDevelopment @deadprogram @trevrosen can any of you guys take a look please? I didn't wanted to ping you, but as there was literally no feedback and Ron didn't answered my email...

**Comment by deadprogram** [comment]
@Shanduur did you see my comment https://github.com/hybridgroup/gocv/pull/720/files#diff-425b7cb72a3bbba55563637d5a320868R876

> // for some reason C.DrawMatches isn't drawing matches

That is not correct, something else must be wrong.

**Comment by shanduur** [comment]
Ok, so side by side analysis of code:

first comparing just the call of `drawMatches` in the `features2d.cpp` file and [OpenCV docs](https://docs.opencv.org/2.4/doc/tutorials/features2d/feature_flann_matcher/feature_flann_matcher.html):

From my pr:
```c++
cv::drawMatches(*img1, kp1vec, *img2, kp2vec, dmatchvec, *outImg,
                cvmatchescolor, cvpointcolor, maskvec,
                static_cast<cv::DrawMatchesFlags>(flags));
```
From example:
```c++
cv::drawMatches(img_1, keypoints_1, img_2, keypoints_2, good_matches, img_matches,
                Scalar::all(-1), Scalar::all(-1), <char>(),
                DrawMatchesFlags::NOT_DRAW_SINGLE_POINTS );
```

As the drawing of keypoints works as it should inside the `drawMatches()` from my PR, we can assume that passing querry, train, and it's corresponding keypoints, as well as the output matrix and colors are passed correctly.

That leads to the assumption that the error is somewhere in the `dmatchvec` or `maskvec`. As the `maskvec` is empty, there is another point that can be eliminated.

So I need to take a deeper look into the `dmatchvec`.

As we can see in the Go code:
```go
for _, d := range matches1to2 {
	p1 := image.Point{
		X: int(kp1[d.QueryIdx].X),
		Y: int(kp1[d.QueryIdx].Y),
	}
	p2 := image.Point{
		X: (int(kp2[d.TrainIdx].X)+img1.Cols()),
		Y: int(kp2[d.TrainIdx].Y),
	}

	Line(outImg, p1, p2, matchColor, 1)
}
```  
We are able to draw lines for matching points.

Next step is to look into the transformation of the `DMatch` array into `vector`:
```c++
for (int i = 0; i < matches1to2.length; i++) {
	dm = cv::DMatch(matches1to2.dmatches[i].queryIdx,
                        matches1to2.dmatches[i].trainIdx,
                        matches1to2.dmatches[i].imgIdx,
                        matches1to2.dmatches[i].distance);
	dmatchvec.push_back(dm);
}
```
And here the creation of this vector is analogous to the creation of `KeyPoint` `vector` few lines above.

So what am I missing?

File: features2d.go [hunk 1]
```diff
@@ -866,27 +865,15 @@ func DrawMatches(img1 Mat, kp1 []KeyPoint, img2 Mat, kp2 []KeyPoint, matches1to2
 	mask := make([]C.char, len(matchesMask))
 
 	cByteArray := C.struct_ByteArray{
-		data:   (*C.char)(&mask[0]),
-		length: (C.int)(len(matchesMask)),
+		length:	(C.int)(len(matchesMask)),
```
Comment (shanduur) [inline features2d.go]: As you can see, I am just dumb...

**Comment by shanduur** [comment]
As you can see, I am just blind and solution was reading OpenCV documentation.

**Comment by shanduur** [comment]
@deadprogram I've added final touches (fixed test) and all seems to be working now as in the plain C++ implementation. I provided wrong mask previously - this is fixed now. Do you have allegations concerning this changes?

File: features2d.go [hunk 2]
```diff
@@ -795,3 +795,85 @@ func (d *SIFT) DetectAndCompute(src Mat, mask Mat) ([]KeyPoint, Mat) {
 
 	return getKeyPoints(ret), desc
 }
+
+// DrawMatches draws matches on combined train and querry images.
+//
+// For further details, please see:
+// https://docs.opencv.org/3.4/d4/d5d/group__features2d__draw.html
```
Comment (deadprogram) [inline features2d.go]: This should be pointing to https://docs.opencv.org/master/d4/d5d/group__features2d__draw.html#gad8f463ccaf0dc6f61083abd8717c261a

**Comment by deadprogram** [comment]
Just two last things. 

- Made comment on the link to the OpenCV help topic for drawMatches.

- This case should probably be covered by adding to the smoke test: https://codecov.io/gh/hybridgroup/gocv/compare/6240320eed51651fa2be9cfd304605b7497f4b6f...4f77450a485ed610f213fdabfed1182e043fd966/diff

Almost there! Thanks for the effort on this @Shanduur

**Comment by shanduur** [comment]
All 4 checks are successful, I think my approach of smoke testing is proper:

> Smoke confirms that it works, and sanity checks that it works according to the specification.

As the test passed, we can confirm that smoke test is written properly?

Thanks for your patience and feedback @deadprogram.

**Comment by deadprogram** [comment]
Thank you for all of the effort on this PR @Shanduur 

Now going to squash/merge into `dev` branch. Thanks again!

**Pull request merged**